[feat] 지역(동네) 추가 기능

pie2457 opened the pull request:
## Issues
- #40 

## What is this PR? 👓
지역(동네) 추가 기능에 대한 PR입니다.

## Key changes 🔑
- 지역(동네) 추가
- 지역(동네) 테스트

## 📖 공부하면서 배웠던 것
### 📌 디미터의 법칙(Law of Demeter)
디미터의 법칙은 “Object-Oriented Programming: An Objective Sense of Style” 에서 처음으로 소개되었다. Demeter라는 프로젝트를 진행하던 개발자들은 어떤 객체가 다른 객체에 대해 지나치게 많이 알다보니, 결합도가 높아지고 좋지 못한 설계를 야기한다는 것을 발견하였다. 그래서 이를 개선하고자 객체에게 자료를 숨기는 대신 함수를 공개하도록 하였는데, 이것이 바로 디미터의 법칙이다. 즉, 디미터의 법칙은 다른 객체가 어떠한 자료를 갖고 있는지 속사정을 몰라야 한다는 것을 의미하며, 이러한 이유로 `Don’t Talk to Strangers(낯선 이에게 말하지 마라)`또는 `Principle of least knowledge(최소 지식 원칙)`으로도 알려져 있다. 또는 직관적으로 이해하기 위해 여러 개의 .(도트)을 사용하지 말라는 법칙으로도 많이 알려져 있으며, 디미터의 법칙을 준수함으로써 캡슐화를 높혀 객체의 자율성과 응집도를 높일 수 있다.
#### ✔️ 객체 지향 프로그래밍과 디미터의 법칙
객체 지향 프로그래밍에서 가장 중요한 것은 "객체가 어떤 데이터를 가지고 있는가?"가 아니라, "객체가 어떤 메세지를 주고 받는가?" 이다. 그렇기에 디미터의 법칙은 객체 지향 프로그래밍에서 상당히 중요한 개념인데, 디미터의 법칙이 위배된다는 것은 올바른 객체 지향 프로그래밍을 하지 못하고 있다는 증거이기도 하다.

💡 디미터 법칙의 기본적인 요약은 다음과 같다.

`클래스 C의 메서드 f는 다음과 같은 객체의 메서드만 호출해야 한다.`
- 클래스 C
- f 가 생성한 객체
- f 메서드에게 전달되는 모든 매개변수
- C 인스턴스 변수에 저장된 객체

pie2457 committed:
style #40 : 메서드명 변경

from, of로 통일하기 위해 변경

pie2457 committed:
feat #40 : 지역(동네) 추가 기능

pie2457 committed:
test #40 : 지역(동네) 추가 기능 테스트

pie2457 committed:
refactor #40 : 검증 부분을 메서드로 분리

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/memberTown/MemberTownService.java:
```diff
@@ -0,0 +1,62 @@
+		Principal principal, AddMemberTownRequest memberTownRequest) {
+		List<MemberTown> memberTowns = memberTownRepository.findAllByMemberId(principal.getMemberId());
+		Optional<MemberTown> duplicated = memberTowns.stream()
+			.filter(memberTown -> memberTown.getRegion().getId().equals(memberTownRequest.getAddressId()))
```
메서드 체이닝이 많네요~ `memberTown`에 해당 로직을 담은 메서드를 하나 만드는건 어떻게 생각하세요?

pie2457 replied:
헐 그 생각은 못해봤어요🥸 감사합니다 공부해보고 적용해볼게요 !!

pie2457 replied:
넵 감사합니다 그렇게 적용해볼게요 !

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/memberTown/MemberTownService.java:
```diff
@@ -0,0 +1,62 @@
+			throw new BadRequestException(ErrorCode.ALREADY_ADDRESS);
+		}
+
+		if (memberTowns.size() > 1) {
```
매직 넘버는 어떤 숫자인지 상수화하면 좋을 것 같아요~

pie2457 replied:
넹 그게 좋겠네요 ! 수정하겠습니당 !

23Yong commented on src/main/java/pie/tomato/tomatomarket/presentation/memberTown/request/AddMemberTownRequest.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
여기서는 dto의 NoArgsConstructor의 접근제어자가 protected 네요 이유가 있나요?

pie2457 replied:
`@RequestBody` 로 Request를 받아오려면 기본생성자가 필요한데 그 외에는 사용하는곳이 없어서 protected로 해놨습니다 !

23Yong reviewed (approved):
수고하셨습니다~
예외사항까지 테스트코드를 작성하신 것이 좋네요

pie2457 committed:
refactor #40 : request들의 accessLevel 설정

pie2457 committed:
refactor #40 : 리뷰사항 반영

- MemberTownService : 매직넘버 상수로 변경
- MemberTown : 서비스에게 MemberTown 객체의 내부를 숨기기 위해 MemberTown에서 검증
